plugins/inventory/inventory.py

```diff
@@ -0,0 +1,162 @@
+    NAME = "inventory"  # used internally by Ansible, it should match the file name but not required
+
+    def read_config_data(self, path):
+        """
+        Reads and validates the inventory source file and environ,
+        storing the provided configuration as options.
+        """
+        with open(path, "r") as inventory_src:
+            cfg = yaml.safe_load(inventory_src)
+        return cfg
+
+    def verify_file(self, path):
+        """
+        return true/false if this is possibly a valid file for this plugin to consume
+        """
+        with open(path, "r") as inventory_src:
+            config_contents = yaml.safe_load(inventory_src)
+        plugin = config_contents.get("plugin")
+        if not plugin:
+            return False
+        if plugin not in [self.NAME, "cloud.terraform.inventory"]:
+            return False
+        return True
+
+    def parse(self, inventory, loader, path, cache=False):
+        super(InventoryModule, self).parse(inventory, loader, path)
+
+        cfg = self.read_config_data(path)  # os.environ is not needed
+
+        terraform_binary = cfg.get("binary_path", None)
+        project_path = cfg.get("project_path", None)
+        state_file = cfg.get("state_file", None)
+
+        if terraform_binary is None:
+            terraform_binary = process.get_bin_path("terraform", required=True)
+
+        if TESTING_STATE_FILE:
+            with open("terraform.tfstateshow", "rb") as state_file_b:
+                state_file = json.loads(state_file_b.read())
+        else:
+            state_file = get_state_content(terraform_binary, project_path, state_file)
+
+        for resource in state_file.get("values", {}).get("root_module", {}).get("resources", {}):
+
+            if resource["type"] == "ansible_group":
+                attributes = resource["values"]
+                inventory.add_group(attributes["group_name"])
+                if attributes["children"]:
+                    for child in attributes["children"]:
+                        inventory.add_group(child)
+                        inventory.add_child(attributes["group_name"], child)
+                if attributes["variables"]:
+                    for key, value in attributes["variables"].items():
+                        inventory.set_variable(attributes["group_name"], key, value)
+
+            if resource["type"] == "ansible_host":
+                attributes = resource["values"]
+                inventory.add_host(attributes["host_name"])
+                if attributes["groups"]:
+                    for group in attributes["groups"]:
+                        inventory.add_group(group)
+                        inventory.add_host(attributes["host_name"], group=group)
+                if attributes["variables"]:
+                    for key, value in attributes["variables"].items():
+                        inventory.set_variable(attributes["host_name"], key, value)
```

Comment:
Extract the mapping logic into a function please. Optimally, you would also separate creating hosts and groups from assigning them to the inventory.

Reply:
Added functions.